### ISSUE TITLE
[feature] Provide a method version of the `exports_sources`

### ISSUE THREAD

**Issue by Minimonium** [conan-io/conan#6474]
Conan-Center-Index uses links to external repositories in the `conandata.yml` and then downloads them in the `source` method. 

The problem is that it creates difficulties for consuming since packages are no longer self-contained. 
External repositories increase the surface for service unavailability. 

In closed environments, It's not enough to use `conan download`/`conan upload` to move a package from the Conan center to an internal remote. One should manually clone the recipe's repository, change the `conandata.yml` to point to a vendored source achieve and export the recipe locally. 

It creates problems for the use of the local cache in out-of-network environments too, although the latest local download caching feature solves it to some extent. 

One possible solution would be storing the download cache in the Artifactory too, but we already have that feature and it's called `exports_sources`. The biggest problem with it is that it can only work with already existing files alongside the recipe. But it's suboptimal to store source files of external projects explicitly alongside the recipe.

I propose that we create an `exports_sources` method alternative, an example of such for the CCI:

```python
def [exports_sources_method](self):
        tools.get(**self.conan_data["sources"][self.version])
        extracted_dir = "Catch2-" + self.version
        os.rename(extracted_dir, self._source_subfolder)
```

It would download and extract the tarball, apply necessary changes to the files, similar to the `source` method. The difference is that it runs on export and automatically stores downloaded files with the recipe.

Related: https://github.com/conan-io/conan/issues/6089

CC @memsharded @markgalpin

**Comment by ytimenkov** [comment]
I was wondering if this could be achieved with some sort of hook which monkey-patches recipe's `source()` and `exports_sources` :).

**Comment by fulara** [comment]
this would solve my problem described here:  https://github.com/conan-io/conan/issues/6356
I could export files based on the version - as that  would be probably available then.

**Comment by memsharded** [comment]
The export_sources() method has been implemented in https://github.com/conan-io/conan/pull/6945, so I think this can be closed now.

A different story is that this method is a valid place for capturing sources from internet origin. We have discussed about this and our take:

- Impossible for ConanCenter. We do ``conan export`` many times to evaluate graphs, etc. If every export does capture sources from, lets say Boost, it is going to take forever.
- export() methods do not have access to settings/options. They are not defined yet at export time, as at that time, only the bare recipe is processed, and profile/settings/options are not even an argument, as the recipe must be common to all configurations

So I think we can close this ticket as implemented, and continue conversation in https://github.com/conan-io/conan/issues/6876. Thanks!
